keps/sig-apps/3715-elastic-indexed-job/README.md

```diff
@@ -0,0 +1,936 @@
+`spec.completionMode` named `ElasticIndexed` 
+
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Introduce a new `spec.completionMode` that allows scaling up and down an indexed Job.
+- Maintain the same success, failure and completion semantics of `Indexed` mode.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Change the existing behavior of `NonIndexed` and `Indexed` modes.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+I have a batch workload that requires two sets of pods: a manager and a group of workers that can scale up or down. 
+The manager communicates with the workers to manage their load; examples of such workloads include distributed HPC 
+and districuted ML training. Examples of frameworks include Horovod with MPI, Spark and Ray.
+
+This workload can be modeled as two Jobs, an `Indexed` Job for the manager and an `ElasticIndexed` Job 
+for the workers. A headless service is created to set up stable hostnames for the worker pods . The workers job is
+scaled up/down by updating `spec.completions` and `spec.parallelism` in tandem.
+
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+There is a concern that changing completion semantics on the fly can be subject to
+race conditions with declaring the job finished. However, after a closer look, this 
+seems to not be an issue because the job controller uses Update when setting the
+finished status on the job object, which is subject to object modified errors if the spec
```

Comment:
Right, but that suggests you're basically hoping that the user is sooner with his change than controller, what if user performs `kubectl edit` and the controller happens to be faster, but since `kubectl edit` issues a patch under the covers both options will be applied, iow. the controller finishing a job and user changing completions. I'd still advise against mutating completions.

Reply:
See above comment about relying on coherent workloads in an autoscaling indexed job that don't complete with success while the job would still need to scale

Reply:
Moving to https://github.com/kubernetes/enhancements/pull/3724#discussion_r1082521252